File: Python3/aoc24/day03/parse.py
"""Module containing the parsing of the input data."""

from pathlib import Path
from re import Pattern, compile

from .instruction import Do, Dont, Instruction, Mul

MUL_INSTRUCTION_REGEX: Pattern[str] = compile(r"mul\((\d+),(\d+)\)")
INSTRUCTION_REGEX: Pattern[str] = compile(r"mul\((\d+),(\d+)\)|do\(\)|don't\(\)")


def parse_input_file(path: Path) -> list[Instruction]:
    """Parse the input file in two lists of integers."""
    instructions: list[Instruction] = []

    for instr in INSTRUCTION_REGEX.finditer(path.read_text()):
        match instr.group(0):
            case "do()":
                instructions.append(Do())
            case "don't()":
                instructions.append(Dont())
            case _:
                instructions.append(Mul(int(instr.group(1)), int(instr.group(2))))

    return instructions

File: Rust/src/day03/parse.rs
use crate::day03::model::Instruction;
use regex::{Captures, Regex};

pub fn parse_input(input: &str) -> Vec<Instruction> {
    let instruction_regex: Regex = Regex::new(r"mul\((\d+),(\d+)\)|do\(\)|don't\(\)").unwrap();
    let mul_instruction_regex: Regex = Regex::new(r"mul\((\d+),(\d+)\)").unwrap();

    let mut result: Vec<Instruction> = Vec::new();

    // Iter on all matches
    for mtc in instruction_regex.find_iter(input) {
        match mtc.as_str() {
            "don't()" => result.push(Instruction::Dont),
            "do()" => result.push(Instruction::Do),
            _ => result.push(parse_mul_instruction(&mul_instruction_regex, mtc.as_str())),
        }
    }

    result
}

fn parse_mul_instruction(mul_regex: &Regex, value: &str) -> Instruction {
    let captures: Captures = mul_regex.captures(value).unwrap();
    let left_val = captures.get(1).unwrap().as_str().parse::<i32>().unwrap();
    let right_val = captures.get(2).unwrap().as_str().parse::<i32>().unwrap();
    Instruction::Mul(left_val, right_val)
}
